conjure_oxide/src/solvers/minion.rs

```diff
@@ -0,0 +1,279 @@
+//! Solver interface to minion_rs.
+
+use crate::ast::{
+    DecisionVariable, Domain as ConjureDomain, Expression as ConjureExpression,
+    Model as ConjureModel, Name as ConjureName, Range as ConjureRange,
+};
+use minion_rs::ast::{
+    Constant as MinionConstant, Constraint as MinionConstraint, Model as MinionModel,
+    Var as MinionVar, VarDomain as MinionDomain,
+};
+
+impl TryFrom<ConjureModel> for MinionModel {
+    // TODO: set this to equal ConjureError once it is merged.
+    type Error = String;
+
+    fn try_from(conjure_model: ConjureModel) -> Result<Self, Self::Error> {
+        let mut minion_model = MinionModel::new();
+
+        // We assume (for now) that the conjure model is fully valid
+        // i.e. type checked and the variables referenced all exist.
+        parse_vars(&conjure_model, &mut minion_model)?;
+        parse_exprs(&conjure_model, &mut minion_model)?;
+
+        Ok(minion_model)
+    }
+}
+
+fn parse_vars(conjure_model: &ConjureModel, minion_model: &mut MinionModel) -> Result<(), String> {
+    // TODO: remove unused vars?
+    // TODO: ensure all vars references are used.
+    for (name, variable) in conjure_model.variables.iter() {
+        parse_var(name, variable, minion_model)?;
+    }
+    Ok(())
+}
+
+fn parse_var(
+    name: &ConjureName,
+    variable: &DecisionVariable,
+    minion_model: &mut MinionModel,
+) -> Result<(), String> {
+    let str_name = name_to_string(name.to_owned());
+
+    // A Conjure-Oxide variable has an arbritary number of ranges as part of its domain. Here, we
+    // require that it has a single range in its domain, and that the domain is of integers.
```

Comment:
this is fine for now. we want to allow arbitrary ranges soon, of course. one way of doing this is to find the full range and then add a constraint to remove the holes.

From:

```
find x:int(1..3, 5..6)
```

To something like:

```
DISCRETE x {1..6}
w-inintervalset(x, [1,3,5,6])
```

Or using sparse-bound in minion. Or both, and letting upstream choose (even longer term).

Reply:
I presume this will be done during tailoring, before we are ready to solve?

Reply:
Sure, we could do it that way.

Reply:
I am not sure what you had in mind, but I presumed that what me and georgii are doing here are the lowest level type conversions, and all other transformations would be done in the more expressive and easier to work with Conjure Model AST.

I imagine the conversion from changing holes in bounds to constraints would be solver independent?

Reply:
it is solver dependent, as in some solvers will natively support it. but this doesn't mean it cannot happen on the conjure-oxide's ast, it will know which solver it's targeting. so I agree it can be done at the higher level.

Reply:
Forgot about sparse bound in minion.

Guess we could have it both ways - if minion gets a model with multiple ranges , it can use sparse bound, but you could alternatively rewrite the ranges as constraints.


Sparse bound will need C++ support, so may not come immediately.

Ill add a Todo in the code to say "put sparse bound here" or something, and we can deal with this higher level before this is implemented.